src/main/java/javaslang/render/text/Box.java

```diff
@@ -154,16 +96,177 @@ public Box connect(Iterable<Box> iterable) {
+        public String toString() {
+            if (contents.length > 0) {
+                StringBuilder builder = new StringBuilder();
+                for (String s : contents)
+                    builder.append("\n").append(s);
+                return builder.substring(1);
+            }
+            return "";
+        }
+
+        private Basic widen(int newWidth, HAlignment alignment) {
+            if (newWidth <= w)
+                return this;
+            String[] result = new String[h];
+            for (int i = 0; i < h; i++)
+                result[i] = justify(contents[i], newWidth, alignment);
+            return new Basic(newWidth, h, result);
+        }
+
+        private Basic heighten(int newHeight, VAlignment alignment) {
+            if (newHeight <= h)
+                return this;
+            String[] result = new String[newHeight];
+            Arrays.fill(result, Box.replicate(' ', w));
+            int topHeight =
+                    alignment == VAlignment.TOP ?
+                            0 :
+                            alignment == VAlignment.CENTER ?
+                                    (newHeight - h) / 2 :
+                                    newHeight - h;
+            System.arraycopy(contents, 0, result, topHeight, h);
+            return new Basic(w, newHeight, result);
+        }
+    }
+
+    public static abstract class Compound extends Box {
+
+        protected Basic[] boxes;
+
+        @Override
+        public String toString() {
+            return toBasic().toString();
+        }
+    }
+
+    public static class Row extends Compound {
+        private VAlignment alignment;
+
+        private Row(VAlignment alignment, Box... contents) {
+            w = 0;
+            h = 0;
+            ArrayList<Basic> list = new ArrayList<>();
+            for (Box b : contents) {
+                h = Math.max(h, b.height());
+                w += b.width();
+                if (b instanceof Row)
+                    list.addAll(Arrays.asList(((Row) b).boxes));
+                else
+                    list.add(b.toBasic());
+            }
+            boxes = list.toArray(new Basic[list.size()]);
+            this.alignment = alignment;
+        }
+
+        @Override
+        protected Basic toBasic() {
+            String[] result = new String[h];
+            Arrays.fill(result, "");
+            for (Basic b : boxes) {
+                String[] s = b.heighten(h, alignment).contents;
+                for (int i = 0; i < h; i++)
+                    result[i] += s[i];
+            }
+            return new Basic(w, h, result);
+        }
+    }
+
+    public static class Col extends Compound {
+        private HAlignment alignment;
+
+        private Col(HAlignment alignment, Box... contents) {
+            w = 0;
+            h = 0;
+            ArrayList<Basic> list = new ArrayList<>();
+            for (Box b : contents) {
+                h += b.height();
+                w = Math.max(w, b.width());
```

Comment:
ah, I see, `w` and `h` actually can't be final. Never mind, they don't have to. It is all right then!